Python/Product/AnacondaInstallLauncher/Program.cs

```diff
@@ -0,0 +1,190 @@
+                    return DoInstall();
+                } else {
+                    return DoUninstall();
+                }
+            } catch (Exception ex) {
+                Console.Error.WriteLine(ex);
+                Console.Error.WriteLine(ex.ToString());
+                return ex.HResult;
+            }
+        }
+
+        void Cancel() => _cancel = true;
+
+        Process Run(string filename, string arguments) {
+            var psi = new ProcessStartInfo(filename, arguments) {
+                CreateNoWindow = true,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false
+            };
+            Console.Error.WriteLine("Starting:{0}  Filename: {1}{0}  Argument: {2}", Environment.NewLine, psi.FileName, psi.Arguments);
+            var p = Process.Start(psi);
+
+            p.OutputDataReceived += (_, e) => { if (e.Data != null) Console.Write(e.Data); };
+            p.ErrorDataReceived += (_, e) => { if (e.Data != null) Console.Error.Write(e.Data); };
+
+            return p;
+        }
+
+        int DoInstall() {
+            WaitForUninstall(false);
+            if (_cancel) {
+                return ERROR_INSTALL_CANCEL;
+            }
+
+            var p = Run(_installer, string.Format(CultureInfo.InvariantCulture, "/InstallationType=AllUsers /RegisterPython=0 /S /D={0}", _targetDir));
+            p.WaitForExit();
+            return p.ExitCode;
+        }
+
+        int DoUninstall() {
+            var p = Run(_installer, "/S");
+
+            if (p.WaitForExit(30000)) {
+                Console.Error.WriteLine("Process exited: {0}", p.ExitCode);
+            } else {
+                Console.Error.WriteLine("Process has not exited");
+            }
+
+            WaitForUninstall(true);
+
+            return _cancel ? ERROR_INSTALL_CANCEL : p.ExitCode;
+        }
+
+        void WaitForUninstall(bool requireWait) {
+            try {
+                WaitForProcess(requireWait);
+            } catch (Exception ex) {
+                Console.Error.WriteLine(ex.ToString());
+            }
+        }
+
+        bool IsDirectoryPresent() {
+            if (!Directory.Exists(_targetDir)) {
+                Console.Error.WriteLine("Target directory is not present");
+                return false;
+            }
+            try {
+                if (!Directory.EnumerateFileSystemEntries(_targetDir).Any()) {
+                    Console.Error.WriteLine("Target directory is empty");
+                    return false;
+                }
+            } catch (Exception ex) {
+                Console.Error.WriteLine("Error reading target directory");
+                Console.Error.WriteLine(ex.ToString());
+                return false;
+            }
+
+            return true;
+        }
+
+        Process[] GetProcesses() {
+            return Process.GetProcesses()
+                .Where(p => p.ProcessName.StartsWith("Un_") &&
```

Comment:
According to the docs at https://msdn.microsoft.com/en-us/library/system.diagnostics.process.processname(v=vs.110).aspx, ProcessName property may throw `InvalidOperationException` if process has exited. Given that we're looking at all processes on the machine, it could happen that a process exits between the call to `GetProcesses` and `ProcessName`.

We print `ProcessName` in `WaitForProcess` as well.